streaming/src/main/scala/org/apache/gearpump/streaming/source/DataSourceTask.scala

```diff
@@ -46,23 +46,39 @@ class DataSourceTask private[source](context: TaskContext, conf: UserConfig, sou
   }
   private val batchSize = conf.getInt(DataSourceConfig.SOURCE_READ_BATCH_SIZE).getOrElse(1000)
   private var startTime = 0L
+  private var lastWatermark: TimeStamp = startTime
 
   override def onStart(newStartTime: StartTime): Unit = {
     startTime = newStartTime.startTime
     LOG.info(s"opening data source at $startTime")
     source.open(context, startTime)
-    self ! Message("start", System.currentTimeMillis())
+    self ! Message("start")
   }
 
   override def onNext(message: Message): Unit = {
     0.until(batchSize).foreach { _ =>
-      Option(source.read()).foreach(context.output)
+      Option(source.read()).foreach { msg =>
+        context.output(msg)
+      }
     }
-    self ! Message("continue", System.currentTimeMillis())
+
+    maybeUpdateWatermark()
+    self ! Message("continue")
   }
 
   override def onStop(): Unit = {
     LOG.info("closing data source...")
     source.close()
   }
+
+  private def maybeUpdateWatermark(): Unit = {
+    val curWatermark = source.getWatermark
+    if (curWatermark > lastWatermark) {
+      lastWatermark = curWatermark
+      self ! UpstreamMinClock(curWatermark)
```

Comment:
Once the watermark move forward, the Actor will send an UpstreamMinClock to self. If DataSource's watermark updates frequently, will it be a problem? I mean the UpstreamMinClock flood.